### PR TITLE
Add import assertions to type only imports and import types to force the resolution mode of the specifier

### LINKED ISSUE #47248: Allow synchronous type-only import of ESM package from CJS
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

import, type, esm

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## 📃 Motivating Example

For a ESM-only package `esm-package`, if we import types from it in a CommonJS file:

```ts
import type {SomeType} from 'esm-package';
```

TypeScript warns in typescript `4.6.0-dev.20211225`:

```
Module 'esm-package' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.ts(1471)
```

For type-only import this shouldn't matter since it doesn't emit JavaScript code, so we can allow 

## 💻 Use Cases

Types, in some cases, are not possible to lazy import asynchronously, such as a top level export:

```ts
import type {SomeType} from 'esm-package';

export const DEFAULT_CONFIG: SomeType = {
  // ...
};
```

### LINKED ISSUE #47338: error importing type from esm in commonjs
# Bug Report

### 🔎 Search Terms

typeof import
import esm type from commonjs module
node12 nodenext

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because `"module": "node12"` only works in `typescript@next`

### ⏯ Playground Link

I was unable to find a way to emulate `"type": "commonjs"` in `package.json` on typescript playground

here is a repo that can be used to reproduce this error
https://github.com/nstringham/min-to-reproduce-typescript-import-problem

### 💻 Code

[/tsconfig.json](https://github.com/nstringham/min-to-reproduce-typescript-import-problem/blob/main/tsconfig.json)
```json
{
  "compilerOptions": {
    "module": "nodenext"
  }
}
```

[/dependency/index.d.ts](https://github.com/nstringham/min-to-reproduce-typescript-import-problem/blob/main/dependency/index.d.ts)
```ts
export type Color = "red" | "green" | "blue";

export function printColor(color: Color): void;
```

[/index.ts](https://github.com/nstringham/min-to-reproduce-typescript-import-problem/blob/main/index.ts)
```ts
import type { Color } from "dependency";

const myColor: Color = "blue";

const myOtherColor: import("dependency").Color = "red";

async function main() {
  const { printColor } = await import("dependency");
  printColor(myColor);
  printColor(myOtherColor);
}

main();
```

### 🙁 Actual behavior

code compiles correctly but displays the following error
```
index.ts:1:28 - error TS1471: Module 'dependency' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.

1 import type { Color } from "dependency";
                             ~~~~~~~~~~~~

index.ts:5:28 - error TS1471: Module 'dependency' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.

5 const myOtherColor: import("dependency").Color = "red";
                             ~~~~~~~~~~~~
```

### 🙂 Expected behavior

code compiles correctly with no errors

### ↪️ Workaround

because typescript compiles correctly despite errors this error can be ignored with `// @ts-ignore`

### PR DESCRIPTION
**Pull request by weswigham**
This PR adds support for a TS-specific assertion on type-only imports and on import type nodes which forces the resolver into either `require` or `import` resolution mode, allowing access to types which the containing file's default mode would normally make difficult or impossible to access. They look like this:
```ts
import type { RequireInterface } from "pkg" assert { "resolution-mode": "require" };
import type { ImportInterface } from "pkg" assert { "resolution-mode": "import" };

export interface LocalInterface extends RequireInterface, ImportInterface {}

export type LocalType =
    & import("pkg", { assert: {"resolution-mode": "require"} }).RequireInterface
    & import("pkg", { assert: {"resolution-mode": "import"} }).ImportInterface;
```

This is essentially a continuation of #47732 that allows resolver configurability for more kinds of type-based imports. Since it's included in this PR, it's *probably* best to go and review & merge that first.

Fixes #47338 by allowing the types of an esm package to be fetched in a cjs file via these modal imports.
Fixes #47248

### CODE REVIEW

File: tests/baselines/reference/nodeModulesImportTypeModeDeclarationEmitErrors1(module=node12).js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+export type LocalInterface =
+    & import("pkg", {"resolution-mode": "require"}).RequireInterface
+    & import("pkg", {"resolution-mode": "import"}).ImportInterface;
+
+export const a = (null as any as import("pkg", {"resolution-mode": "require"}).RequireInterface);
+export const b = (null as any as import("pkg", {"resolution-mode": "import"}).ImportInterface);
+//// [other2.ts]
+// wrong assertion key
+export type LocalInterface =
+    & import("pkg", { assert: {"resolution-m0de": "require"} }).RequireInterface
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/nodeModulesImportTypeModeDeclarationEmitErrors1(module=node12).js:34]: Let's make this more obviously not-the-thing

File: src/compiler/program.ts [hunk 2]
```diff
@@ -553,9 +553,39 @@ namespace ts {
...
+        if (isExportDeclaration(decl)) {
+            return decl.isTypeOnly;
+        }
+        if (decl.importClause?.isTypeOnly) {
+            return true;
+        }
+        const bindings = decl.importClause?.namedBindings;
+        if (bindings && isNamedImports(bindings) && length(bindings.elements) && every(bindings.elements, b => b.isTypeOnly)) {
+            return true;
+        }
```
Comment (andrewbranch) [inline src/compiler/program.ts:567]: This shouldn’t be considered here—at the very least, you need to check that `compilerOptions.importsNotUsedAsValues` is not `preserve` or `error`, but I think the more coherent solution is that only `import type ...` is allowed to specify an override. A regular import whose specifiers happen to all be marked type-only is really no different than a regular import who happens to import only types. I want to reinforce the mental model that `import type ...` → no emit (therefore can have special behavior) and `import { type Foo }` → `import {}`. If the latter gets elided from JS emit, that’s a follow-on effect of our unusually aggressive import elision; it doesn’t mean the source import was special.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -8693,6 +8708,7 @@ namespace ts {
         readonly includePackageJsonAutoImports?: "auto" | "on" | "off";
         readonly provideRefactorNotApplicableReason?: boolean;
         readonly jsxAttributeCompletionStyle?: "auto" | "braces" | "none";
+        readonly overrideImportMode?: false | "require" | "import";
```
Comment (andrewbranch) [inline src/compiler/types.ts:8711]: I don’t think this makes sense as a “user preference”—seems like it should just be another parameter to `getModuleSpecifiers`. These all have user-facing analogs in protocol.ts and settings in VS Code. It’s kind of a coincidence that a couple of them are useful for declaration emit.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
This feels really wrong :confused: 

The proposal explicitly says that [import assertions must not affect how a module is evaluated or resolved](https://tc39.es/proposal-import-assertions/#sec-hostresolveimportedmodule):
> [...]
> The implementation of HostResolveImportedModule must conform to the following requirements:
> - [...]
> - moduleRequest.[[Assertions]] must not influence the interpretation of the module or the module specifier; instead, it may be used to determine whether the algorithm completes normally or with an [abrupt completion](https://tc39.es/ecma262/#sec-completion-record-specification-type).

In other words, if both these two imports succeed they must resolve to the same file:
```js
import { A } from "pkg" assert { "resolution-mode": "require" };
import { A } from "pkg" assert { "resolution-mode": "import" };
```

The imports assertions proposal [even mentions](https://github.com/tc39/proposal-import-assertions#follow-up-proposal-evaluator-attributes) a possible follow-up proposal that would allow what you need, but it would be a separate proposal (with likely different syntax).

I know that this PR is only about `import type` so it's technically "just TS" and not JS, but actively going against the JS semantics should be discouraged.

**Comment by nicolo-ribaudo** [comment]
https://github.com/tc39/proposal-import-reflection is probably what you need.

**Comment by weswigham** [comment]
These don't affect runtime and are type-only; it's fine - it's much better than introducing full on new syntax.

**Comment by nayeemrmn** [comment]
> it's fine - it's much better than introducing full on new syntax.

@weswigham This is wrong, we're using the `assert` keyword for something that isn't an assertion. Why? You're correct that this is a type-only statement, all the more reason to do literally anything else and not misuse a JS syntax.

There are plenty of other options that make sense without _completely_ new and ts-only syntax e.g.
```ts
// comment directive, nothing new

// @ts-resolution=require
import type { RequireInterface } from "pkg";
```
```ts
// https://github.com/tc39/proposal-import-reflection

import type { RequireInterface } from "pkg" as "resolution=require";
```
```ts
// dangles off the already ts-only part

import type(resolution=require) { RequireInterface } from "pkg";
```

Please consider reverting this change.

**Comment by nayeemrmn** [comment]
I've already seen a lot of people confused by import assertions and what they could be used for in runtimes, specifically conflating them with what's proposed in https://github.com/tc39/proposal-import-reflection. It would be a shame to see TS fostering that misconception, or perhaps falling victim to it.

**Comment by glen-84** [comment]
Why is it:

```ts
assert {"resolution-mode": "import"};
```

(requiring quotes), and not:

```ts
assert {resolutionMode: "import"};
```

?

**Comment by weswigham** [comment]
Consistency with the triple-slash reference form and a bit of extra syntax encumbrance to make you pause and think if you _really_ need to use it. :)

**Comment by frehner** [comment]
Would this also hypothetically help solve the issue found at https://github.com/microsoft/TypeScript/issues/46213 ? If so, how soon do you think this will graduate from nightly?

**Pull request merged**